apigee/registry
# Implement Spectral Linter Plugin according to new Plugin Architecture

## Summary
The following pull request implements the spectral linter plugin and follows the new plugin architecture that was recently established in the registry. This work was previously implemented in . For a detailed description, please read the description of that PR.

## Test
- Have registry running on `localhost:8080`.
- Ensure that https://github.com/APIs-guru/openapi-directory is extracted to your Desktop.
- Ensure that the following style guide is located under `~/styleguide.yaml`:
```yaml
id: openapistyleguide
mime_types:
  - application/x.openapi+gzip;version=2
  - application/x.openapi+gzip;version=3
guidelines:
  - id: refproperties
    display_name: Govern Ref Properties
    description: This guideline governs properties for ref fields on specs.
    rules:
      - id: norefsiblings
        description: An object exposing a $ref property cannot be further extended with additional properties.
        linter: spectral
        linter_rulename: no-$ref-siblings
        severity: 1
    status: 2
linters:
  - name: spectral
    uri: https://github.com/stoplightio/spectral
```
- Compile the registry project by running `make` in the root directory
- Change directory to `$GOPATH/bin`
- Now, run the following series of bash commands:
```bash
apg registry delete-project --name projects/example

apg registry create-project --project_id example --insecure --address localhost:8080

registry upload bulk openapi --project-id example ~/Desktop/openapi-directory/APIs

registry upload styleguide ~/styleguide.yaml --project-id=example

registry compute conformance projects/example/locations/global/apis/billbee.io/versions/v1/specs/swagger.yaml --plugin=true

registry get projects/example/locations/global/apis/billbee.io/versions/v1/specs/swagger.yaml/artifacts/conformance-openapistyleguide --contents
```
Ensure that the output matches the [following gist](https://gist.github.com/muhammadharis/346c9818ab24f1156a380028b8f25da9).
